File: src/main/java/leetcode/lists/topinterview/LeetCode277.java
package leetcode.lists.topinterview;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/10/5 13:19
 */
public class LeetCode277 {


    private int[][] graph;

    public int findCelebrity(int n) {
        int ans = 0;
        for (int k = 0; k < n; k++) {
            if (knows(ans, k)) {
                ans = k;
            }
        }

        for (int k = 0; k < n; k++) {
            if (k != ans) {
                if (!knows(k, ans)) {
                    return -1;
                }
                if (knows(ans, k)) {
                    return -1;
                }
            }
        }
        return ans;
    }


    //获取到 A 是否认识 B。
    private boolean knows(int i, int j) {
        return graph[i][j] == 1;
    }
}

File: src/main/go/leetcode/lists/topinterview/LeetCode277.go
package topinterview

func solution(knows func(a int, b int) bool) func(n int) int {
	return func(n int) int {
		celebrity := 0
		for i := 1; i < n; i++ {
			if knows(celebrity, i) {
				celebrity = i
			}
		}
		for i := 0; i < n; i++ {
			if i == celebrity {
				continue
			}
			if knows(celebrity, i) || !knows(i, celebrity) {
				return -1
			}
		}
		return celebrity
	}
}
